### ISSUE TITLE
Email Label Suggestion via LLM mini model (deferred from #61 ML training pipeline)

### ISSUE THREAD

**Issue by mauricio-morales** [mauricio-morales/smart-inbox-janitor#77]
## Overview

This issue captures the full design for a local ML.NET label prediction model that was scoped out of feature #61 (ML.NET Training Pipeline). The approach has been reconsidered: instead of training a local OvA multi-label classifier, future label suggestion will be driven by an **OpenAI mini model** (e.g. `gpt-4o-mini`) that receives email content and returns label recommendations — making suggestions more useful for users who do not have rich historical labeling data.

This issue preserves all design research from the original scoping so nothing is lost if we revisit the local ML-based approach in the future.

---

## Why Deferred

Local multi-label classification (OvA binary classifiers per label) requires a user to have rich historical label data across many distinct Gmail labels. Most users don't apply labels heavily, so there is not enough training signal to build a useful model. An OpenAI mini model can infer appropriate labels from email content without any historical data, producing more useful results immediately.

---

## User Story (from spec #61)

> A user wants the system to automatically suggest which Gmail labels should be applied to incoming emails, based on how they historically labeled similar messages. They train a label prediction model that can recommend multiple labels per email with individual confidence scores.

**Acceptance Scenarios:**
1. Given labeled email feature vectors with associated Gmail label history, When the label model is trained, Then the system produces a multi-label classifier that can predict multiple labels per email, each with an individual confidence score
2. Given an email classified by the label model, When retrieving predictions, Then labels are returned in descending confidence order and the caller can request the top-k most likely labels (e.g. top 3)
3. Given some labels are rare in the training corpus, When the model trains, Then class imbalance handling ensures rare labels are still evaluated and not silently ignored
4. Given a minimum confidence threshold is configured per label, When the model makes predictions, Then only predictions meeting or exceeding each label's threshold are surfaced, preventing low-confidence noise
5. Given an email whose content resembles a label the model has never seen, When classification runs, Then the model gracefully returns no prediction for that label rather than a spurious assignment

**Edge cases:**
- If the label model is asked to predict labels for an email from a source with no historical label patterns → return empty prediction list with zero-confidence indicator
- If top-k exceeds the number of known labels → return all known labels ranked by confidence without error

---

## Functional Requirements (from spec #61, FR-006–FR-010, FR-018)

- **FR-006**: System MUST provide a label classification model that predicts zero or more Gmail labels per email, each with an individual confidence score
- **FR-007**: System MUST support configurable minimum confidence thresholds per label so low-confidence predictions are suppressed
- **FR-008**: System MUST support top-k label retrieval, returning only the k highest-confidence label predictions for a given email
- **FR-009**: System MUST handle class imbalance in label training data so rare labels are evaluated and not silently suppressed
- **FR-010**: System MUST gracefully return no prediction (empty set, not an error) for labels the model has not encountered in training
- **FR-018**: System MUST maintain a separate version history for the label model (distinct from the action model version history)

**Former success criteria:**
- **SC-002**: With 500+ labeled emails, label model training completes in under 2 minutes and produces predictions where top-1 label precision ≥ 0.75 on held-out validation data
- **SC-008**: The system correctly suppresses low-confidence label predictions at configurable thresholds, reducing false positive label assignments by at least 30% compared to unconstrained prediction

---

### R2: ML.NET Trainer Selection — Label Model

**Original decision**: One-vs-All (OvA) binary `LbfgsLogisticRegression` classifier per label.

**Rationale:**
- ML.NET does not provide a native multi-label classifier. The standard approach is to train one binary classifier per label, treating each as an independent binary prediction problem.
- `LbfgsLogisticRegression` with LBFGS converges quickly for binary logistic regression and produces well-calibrated probabilities, cleanly supporting per-label confidence thresholds.
- Training N binary classifiers (one per label) is embarrassingly parallelisable with `Parallel.ForEach`, fitting within the <2 min target for up to 50 labels.
- Each classifier's probability output is independently thresholded (`LabelThresholds` in config), enabling per-label suppression of low-confidence predictions.

**Alternatives considered:**
- `SdcaMaximumEntropy` via `OneVersusAll` wrapper: Available in ML.NET but the `OneVersusAll` meta-estimator doesn't expose per-label probability thresholds cleanly.
- Deep-learning multi-label head (TorchSharp): Unnecessarily complex; violates the offline/all-.NET constraint.

### R3: Class Balancing for Label Model

For the OvA label model, a minimum of 5 positive examples per label would be enforced before a binary classifier is trained. Labels with fewer examples would be skipped with a diagnostic warning rather than producing a degenerate model. Inverse-frequency row weights via `ExampleWeightColumnName` handle class imbalance within each binary classifier.

---

### ILabelClassifier

```csharp
public interface ILabelClassifier
{
    bool IsLoaded { get; }
    string? LoadedModelId { get; }
    IReadOnlySet<string> KnownLabels { get; }

    Task<Result<bool>> LoadModelAsync(string modelFilePath, CancellationToken ct = default);
    void UnloadModel();

    Result<IReadOnlyList<LabelPrediction>> PredictLabels(EmailFeatureVector input, int topK = 3);
    Result<IReadOnlyList<IReadOnlyList<LabelPrediction>>> PredictLabelsBatch(
        IEnumerable<EmailFeatureVector> inputs, int topK = 3);
    float GetThreshold(string labelName);
}
```

### IMLModelProvider label methods

```csharp
Task<Result<IReadOnlyList<LabelPrediction>>> PredictLabelsAsync(
    EmailFeatureVector input, int topK = 3, CancellationToken ct = default);

Task<Result<IReadOnlyList<IReadOnlyList<LabelPrediction>>>> PredictLabelsBatchAsync(
    IEnumerable<EmailFeatureVector> inputs, int topK = 3, CancellationToken ct = default);
```

### IModelTrainingPipeline label methods

```csharp
Task<Result<TrainingMetricsReport>> TrainLabelModelAsync(
    TrainingRequest request,
    IProgress<TrainingProgress>? progress = null,
    CancellationToken ct = default);

Task<Result<TrainingMetricsReport>> IncrementalUpdateLabelModelAsync(
    IncrementalUpdateRequest request,
    IProgress<TrainingProgress>? progress = null,
    CancellationToken ct = default);

Task<Result<TrainingDataSummary>> GetLabelTrainingDataSummaryAsync(CancellationToken ct = default);
```

### ML.NET Data Types

**LabelTrainingInput** — per-label binary training input schema for ML.NET IDataView:

```csharp
public class LabelTrainingInput
{
    public float SenderKnown { get; set; }
    public float ContactStrength { get; set; }
    public float HasListUnsubscribe { get; set; }
    public float HasAttachments { get; set; }
    public float HourReceived { get; set; }
    public float DayOfWeek { get; set; }
    public float EmailSizeLog { get; set; }
    public float SubjectLength { get; set; }
    public float RecipientCount { get; set; }
    public float IsReply { get; set; }
    public float LabelCount { get; set; }
    public float EmailAgeDays { get; set; }
    public float IsInInbox { get; set; }
    public float IsStarred { get; set; }
    public float IsArchived { get; set; }
    public float SenderFrequency { get; set; }
    public string SenderDomain { get; set; } = string.Empty;  // categorical
    public string SubjectText { get; set; } = string.Empty;   // TF-IDF
    public float Weight { get; set; } = 1.0f;
    public bool BinaryLabel { get; set; }  // true = email had this label
}
```

**LabelPrediction** — output of one OvA binary classifier:

```csharp
public class LabelPrediction
{
    public string LabelName { get; set; } = string.Empty;
    public float Confidence { get; set; }
    public bool MeetsThreshold { get; set; }
}
```

---

## Preferred Future Approach: OpenAI Mini Model for Label Suggestions

Rather than training a local ML model, label suggestions should be obtained from a lightweight LLM:

- **Model**: `gpt-4o-mini` (or equivalent low-cost mini model)
- **Prompt strategy**: Provide the email subject, sender, and first ~500 chars of body; supply the user's existing Gmail label names as a constrained choice set; ask the model to suggest 0–3 labels with brief reasoning
- **Integration point**: Extend `ILLMProvider` (or create a new `ILabelSuggestionProvider`) with a `SuggestLabelsAsync(EmailSummary email, IReadOnlyList<string> availableLabels)` method returning `Result<IReadOnlyList<LabelSuggestion>>`
- **Cost guard**: Label suggestion is opt-in per session (not auto-run on every email) to control API costs
- **Offline mode**: Falls back gracefully to no suggestions when the LLM provider is unavailable or not configured

---

## Source Files Removed from Plan #61

These files were designed during the planning of #61 but removed from scope when label prediction was deferred:

- `specs/059-mlnet-training-pipeline/contracts/ILabelClassifier.cs` (deleted)
- Label methods removed from `IMLModelProvider` and `IModelTrainingPipeline` contracts
- `LabelTrainingInput`, `LabelPrediction` types removed from data-model.md
- `LabelModelTrainer.cs`, `LabelClassifier.cs`, `LabelTrainingInput.cs`, `LabelPrediction.cs` removed from implementation plan
- Label-specific success criteria SC-002 and SC-008 removed from spec
- `IncrementalUpdateLabelModelAsync`, `TrainLabelModelAsync`, `GetLabelTrainingDataSummaryAsync` removed from IModelTrainingPipeline

---

## Dependencies for Revisiting Local ML Approach

- #54 (ML architecture design) — feature definitions
- #55 (ML data storage) — `EmailFeatureVector` including label history fields
- #61 (ML.NET training pipeline) — action model infrastructure (reusable)
- Sufficient labeled email data: minimum 100 emails with at least 5 examples per target label
